File: day_01/ex01.py
# **************************************************************************** #
#                                                                              #
#                                                         :::      ::::::::    #
#    ex01.py                                            :+:      :+:    :+:    #
#                                                     +:+ +:+         +:+      #
#    By: amenses- <amenses-@student.42porto.com>    +#+  +:+       +#+         #
#                                                 +#+#+#+#+#+   +#+            #
#    Created: 2023/12/01 15:25:54 by amenses-          #+#    #+#              #
#    Updated: 2023/12/02 12:45:00 by amenses-         ###   ########.fr        #
#                                                                              #
# **************************************************************************** #

# Day 01: Trebuchet?
# Part One: Find the first and last numbers in each line and add them together.

with open("input.txt", "r") as f:
    result = 0
    for line in f:
        first_number = [char for char in line if char.isdigit()][0]
        last_number = [char for char in line if char.isdigit()][-1]
        result += int(first_number + last_number)
print(result)

# SOLUTION: 54239

File: day_08/ex01.cpp
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   ex01.cpp                                           :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: amenses- <amenses-@student.42porto.com>    +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2023/12/08 14:28:46 by amenses-          #+#    #+#             */
/*   Updated: 2023/12/12 18:20:11 by amenses-         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

/* Day 08 - Part One */

#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <map>
#include <sstream>
#include <unordered_map>
#include <stdio.h>

using namespace std;

int main()
{
    // read input
    ifstream                                        file("input.txt");
    string                                          instructions;
    unordered_map<string, pair<string, string> >    nodes;

    getline(file, instructions, '\n');
    while (file.good())
    {
        string line;

        getline(file, line, '\n');
        if (line.length() == 0)
            continue ;

        istringstream iss(line);
        string node, left, right;
        
        iss >> node >> left >> left >> right;
        nodes[node] = make_pair(left.substr(1, 3), right.substr(0, 3));
    }
    file.close();

    unordered_map<string, pair<string, string> >::iterator  current = nodes.find("AAA");
    int                                                     result = 0;

    while (current->first != "ZZZ")
    {
        for (size_t i = 0; i < instructions.length(); i++)
        {
            if (instructions[i] == 'L')
                current = nodes.find(current->second.first);
            else
                current = nodes.find(current->second.second);
            result++;
            if (current->first == "ZZZ")
                break ;
        }
    }
    cout << result << endl;
    return 0;
}
